src/PostgREST/Request/QueryParams.hs

```diff
@@ -0,0 +1,484 @@
+--
+-- The canonical representation of the query string has paramters sorted alphabetically:
+--
+-- >>> qsCanonical <$> parse "a=1&c=3&b=2&d"
+-- Right "a=1&b=2&c=3&d="
+--
+-- 'select' is a reserved parameter that selects the fields to be returned:
+--
+-- >>> qsSelect <$> parse "select=name,location"
+-- Right [Node {rootLabel = (("name",[]),Nothing,Nothing,Nothing,Nothing), subForest = []},Node {rootLabel = (("location",[]),Nothing,Nothing,Nothing,Nothing), subForest = []}]
+--
+-- Filters are parameters whose value contains an operator, separated by a '.' from its value:
+--
+-- >>> qsFilters <$> parse "a.b=eq.0"
+-- Right [(["a"],Filter {field = ("b",[]), opExpr = OpExpr False (Op OpEqual "0")})]
+--
+-- If the operator specified in a filter does not exist, parsing the query string fails:
+--
+-- >>> qsFilters <$> parse "a.b=noop.0"
+-- Left (QueryParamError "\"failed to parse filter (noop.0)\" (line 1, column 6)" "unknown single value operator noop")
+parse :: ByteString -> Either QueryParamError QueryParams
+parse qs =
+  QueryParams
+    canonical
+    params
+    ranges
+    <$> pRequestOrder `traverse` order
+    <*> pRequestLogicTree `traverse` logic
+    <*> pRequestColumns columns
+    <*> pRequestSelect select
+    <*> pRequestFilter `traverse` filters
+    <*> pRequestFilter `traverse` filtersRoot
+    <*> pRequestFilter `traverse` filtersNotRoot
+    <*> pure (S.fromList (fst <$> filters))
+    <*> sequenceA (pRequestOnConflict <$> onConflict)
+  where
+    logic = filter (endingIn ["and", "or"] . fst) nonemptyParams
+    select = fromMaybe "*" $ lookupParam "select"
+    onConflict = lookupParam "on_conflict"
+    columns = lookupParam "columns"
+    order = filter (endingIn ["order"] . fst) nonemptyParams
+    limits = filter (endingIn ["limit"] . fst) nonemptyParams
+    offsets = first (replaceLast "limit") <$> filter (endingIn ["offset"] . fst) nonemptyParams
+    lookupParam :: Text -> Maybe Text
+    lookupParam needle = toS <$> join (L.lookup needle qParams)
+    nonemptyParams = mapMaybe (\(k, v) -> (k,) <$> v) qParams
+
+    qString = HTTP.parseQueryReplacePlus True qs
+
+    qParams = [(T.decodeUtf8 k, T.decodeUtf8 <$> v)|(k,v) <- qString]
+
+    canonical =
+      BS.pack $ HTTP.urlEncodeVars
+        . L.sortOn fst
+        . map (join (***) BS.unpack . second (fromMaybe mempty))
+        $ qString
+
+    endingIn:: [Text] -> Text -> Bool
+    endingIn xx key = lastWord `elem` xx
+      where lastWord = L.last $ T.split (== '.') key
+
+    (filters, params) = L.partition isParam filtersAndParams
+    isParam (k, v) = isEmbedPath k || hasOperator v || hasFtsOperator v
+
+    filtersAndParams = filter (isFilterOrParam . fst)  nonemptyParams
+    isFilterOrParam k = not (endingIn reservedEmbeddable k) && notElem k reserved
+    reserved = ["select", "columns"]
+    -- TODO: check potential previous bug, on_conflict was missing
+    reservedEmbeddable = ["order", "limit", "offset", "and", "or", "on_conflict"]
```

Comment:
`on_conflict` was missing orginally - not sure if that would have created issues previously, but I don't think it hurts to add it here

Reply:
Should this not be added to `reserved` one line above that instead?

Reply:
yes, good point - it does not apply to embedded fields. Fixed!

Reply:
Actually.. I think it might not be a good idea to add this here at all. I didn't test that, yet, but I understand that part of the code to block any RPC from using an argument by the name of `on_conflict`. Since this is not strictly needed, because `on_conflict` is only a thing for `POST` to tables/views, it would actually be a breaking change to add this here - in case somebody is using such an argument name.

If that holds true, it's actually the other way around: I think we can relax things a little bit and not reserve `columns` - because those are not used when calling RPCs either... right?

Reply:
I think it's quite confusing to only have the parameters be reserved in special cases... But let's not change the behavior here, I've added logic to handle the current 'on_conflict' behavior explicitly - let's see if we want to keep it in a separate PR.

Reply:
I just tested - and `POST`- upsert does not even support filters at all. This is because `POST` will never add a `WHERE` clause:

https://github.com/PostgREST/postgrest/blob/74f6e5225cc2d4c3b8b13bb99a80224592240644/src/PostgREST/Query/QueryBuilder.hs#L89-L90

So this is actually a non-issue. I think you shouldn't add that logic here.